src/components/Tree/Tree.tsx

```diff
@@ -64,40 +64,107 @@ const reducer = (state: TreeState, action: TreeStateAction): TreeState => {
+                const newArray = [
+                    ...arrayWithoutElement.slice(0, targetIndexInParent + indexOffset),
+                    element,
+                    ...arrayWithoutElement.slice(targetIndexInParent + indexOffset),
+                ];
+
+                state.items.set(targetState.parentId, { ...parentTargetState, childrenIds: newArray });
+            } else if (position === DropZonePosition.Within) {
+                const sourceState = state.items.get(id);
+                if (!sourceState || !sourceState.parentId) {
+                    throw new Error(`No tree item registered with id ${id ?? sourceState?.parentId}.`);
+                }
+                const targetState = state.items.get(targetId);
+                const oldParentState = state.items.get(sourceState.parentId);
+                if (!targetState || !oldParentState) {
+                    throw new Error(`No tree item registered with id ${targetId ?? sourceState.parentId}.`);
+                }
+
+                const oldParentChildrenIds = oldParentState.childrenIds?.filter((childId) => childId !== id);
+
+                state.items
+                    // Delete from old parent
+                    .set(sourceState.parentId, {
+                        ...oldParentState,
+                        childrenIds:
+                            oldParentChildrenIds && oldParentChildrenIds.length > 0 ? oldParentChildrenIds : undefined,
+                    })
+                    // Add to new parent
+                    .set(targetId, { ...targetState, childrenIds: [...(targetState?.childrenIds ?? [], id)] })
+                    // Set new parent in source
+                    .set(id, { ...sourceState, parentId: targetId });
+            }
+
+            return {
+                ...state,
+            };
+        }
+
+        case 'SET_SELECTION_MODE': {
             return {
                 ...state,
                 selectionMode: action.payload.selectionMode,
             };
+        }
 
         case 'REGISTER_TREE_ITEM': {
-            const { id, childrenIds, level, parentId, domElement: ref } = action.payload;
+            const { id, level, parentId, domElement: ref } = action.payload;
 
             if (state.items.has(id)) {
                 console.warn('Tree item already registered.');
                 return state;
             }
 
-            if (parentId) {
-                const parentState = state.items.get(parentId);
-                if (parentState) {
-                    const newChildrenIds = [...(parentState?.childrenIds ?? []), id];
-                    state.items.set(parentId, { ...parentState, childrenIds: newChildrenIds });
-                }
-            }
-
             return {
                 ...state,
                 items: state.items.set(id, {
                     parentId,
-                    childrenIds,
                     level,
                     domElement: ref,
+                    childrenIds: undefined,
```

Comment:
why not pass an empty array? So you don't have to do `?` checks

Reply:
I prefer `undefined` for the leafs of a tree